### PR TITLE
7578 Fix/improve some aspects of ZIL writing.

### PR DESCRIPTION
**Pull request by amotin**
- After some ZIL changes 6 years ago zil_slog_limit got partially broken
due to zl_itx_list_sz not updated when async itx'es upgraded to sync.
Actually because of other changes about that time zl_itx_list_sz is not
really required to implement the functionality, so this patch removes
some unneeded broken code and variables.

 - Original idea of zil_slog_limit was to reduce chance of SLOG abuse by
single heavy logger, that increased latency for other (more latency critical)
loggers, by pushing heavy log out into the main pool instead of SLOG.  Beside
huge latency increase for heavy writers, this implementation caused double
write of all data, since the log records were explicitly prepared for SLOG.
Since we now have I/O scheduler, I've found it can be much more efficient
to reduce priority of heavy logger SLOG writes from ZIO_PRIORITY_SYNC_WRITE
to ZIO_PRIORITY_ASYNC_WRITE, while still leave them on SLOG.

 - Existing ZIL implementation had problem with space efficiency when it
has to write large chunks of data into log blocks of limited size.  In some
cases efficiency stopped to almost as low as 50%.  In case of ZIL stored on
spinning rust, that also reduced log write speed in half, since head had to
uselessly fly over allocated but not written areas.  This change improves
the situation by offloading problematic operations from z*_log_write() to
zil_lwb_commit(), which knows real situation of log blocks allocation and
can split large requests into pieces much more efficiently.  Also as side
effect it removes one of two data copy operations done by ZIL code WR_COPIED
case.

 - While there, untangle and unify code of z*_log_write() functions.
Also zfs_log_write() alike to zvol_log_write() can now handle writes crossing
block boundary, that may also improve efficiency if ZPL is made to do that.

Sponsored by:   iXsystems, Inc.

### CODE REVIEW

**Comment by prakashsurya** [comment]
@zettabot go

**Comment by ahrens** [comment]
Lint warning will need to be fixed:
`"../../common/fs/zfs/zil.c", line 1382: warning: set but not used in function: push_sod in zil_get_commit_list (E_FUNC_SET_NOT_USED)`

**Comment by amotin** [comment]
Removed the push_sod remnants.  Haven't noticed it while built/tested on FreeBSD.  Thanks for pointing.

Could somebody rerun zettabot again?

**Comment by ahrens** [comment]
@zettabot go

**Comment by amotin** [comment]
One more fallout in ztest.  Fixed and rebased.

**Comment by prakashsurya** [comment]
@zettabot go

**Comment by amotin** [comment]
Any idea what does "Agent went offline during the build" means?

**Comment by prakashsurya** [comment]
> Any idea what does "Agent went offline during the build" means?

Yes, that generally means the Jenkins master lost connectivity with the slave's process that runs on the client where the test is being performed. I could give more details about how the tests are run, but I'm not sure that's necessary. Usually this occurs due to the system performing the test rebooting due to a kernel panic, and that appears to be the case here.

Logging into the system that was running `ztest`, I see the uptime is:
```
# uptime
 20:41pm  up   1:10,  1 user,  load average: 0.11, 0.25, 0.64
```
whereas it should be at least 2 hours since the first log line from the zloop output shows a time of `18:10:24`.

I also see a core file in `/var/crash`:
```
# cd /var/crash/unknown/
# ls
bounds  vmdump.0

# savecore -f vmdump.0
...
 1:25 100% done: 441734 of 441734 pages saved
 
# mdb 0
Loading modules: [ unix genunix specfs dtrace mac cpu.generic uppc pcplusmp scsi_vhci zfs
sbd mm idm cpc random lofs ufs logindmux ptm nfs ]
> ::status
debugging crash dump vmcore.0 (64-bit) from omnios
operating system: 5.11 origin/pr/219/head-0-gd1c7d63 (i86pc)
image uuid: a3d367f6-9cb5-c606-e0c5-e72a655e1236
panic message: assertion failed: size != 0, file: ../../common/fs/zfs/zfs_vnops.c, line: 1
dump content: kernel pages only
> ::stack
vpanic()
0xfffffffffbe10cd8()
zfs_get_data+0xe8(ffffff0713991000, ffffff0717182f40, ffffff0717183000, ffffff0bc7896888)
zil_lwb_commit+0x276(ffffff070e20d780, ffffff072f2d7640, ffffff074119eb68)
zil_commit_writer+0xb3(ffffff070e20d780)
zil_commit+0x9e(ffffff070e20d780, 70741)
zfs_fsync+0xba(ffffff0722e65500, 10, ffffff072d7b0d60, 0)
fop_fsync+0x44(ffffff0722e65500, 10, ffffff072d7b0d60, 0)
fdsync+0x3c(8, 10)
sys_syscall+0x26e()
```
So, while running `ztest` the system panicked with the above stack trace.

**Comment by amotin** [comment]
I am still not sure how that panic happened exactly, but I've fixed different bug that could crash on assert in different but alike place.

**Comment by amotin** [comment]
Thinking again, my last change could indeed fix the problem. Please restart the test, if not difficult, while I also run more tests on my own.  Thanks.

**Comment by ahrens** [comment]
@zettabot go

**Comment by amotin** [comment]
Seems like it no longer crashes.  Good.  But it failed bunch of tests that I can't imagine related to ZIL. Is that just a backgroud noise?

**Comment by prakashsurya** [comment]
@amotin yes, those are known failures; adding the "test pass" label.

File: usr/src/uts/common/fs/zfs/zil.c [hunk 1]
```diff
@@ -1077,25 +1082,30 @@ zil_lwb_commit(zilog_t *zilog, itx_t *itx, lwb_t *lwb)
...
 
+cont:
 	/*
 	 * If this record won't fit in the current log block, start a new one.
+	 * For WR_NEED_COPY optimize layout for minimal number of chunks, but
+	 * try to keep wasted space withing reasonable range (12%).
 	 */
-	if (lwb->lwb_nused + reclen + dlen > lwb->lwb_sz) {
+	lwb_sp = lwb->lwb_sz - lwb->lwb_nused;
+	if (reclen > lwb_sp || (reclen + dlen > lwb_sp &&
```
Comment (prakashsurya) [inline usr/src/uts/common/fs/zfs/zil.c]: Even with the comment above, the purpose of these additional checks in this conditional still isn't clear to me. Could you try to add some more info to describe why we want these additional checks (i.e. this line, and the two below)?

I'm still a little new to the details of the ZIL, so perhaps my inexperience is showing.

File: usr/src/uts/common/fs/zfs/zil.c [hunk 2]
```diff
@@ -1057,12 +1061,13 @@ zil_lwb_write_start(zilog_t *zilog, lwb_t *lwb)
 static lwb_t *
 zil_lwb_commit(zilog_t *zilog, itx_t *itx, lwb_t *lwb)
 {
-	lr_t *lrc = &itx->itx_lr; /* common log record */
-	lr_write_t *lrw = (lr_write_t *)lrc;
+	lr_t *lrcb, *lrc = &itx->itx_lr; /* common log record */
```
Comment (prakashsurya) [inline usr/src/uts/common/fs/zfs/zil.c]: Why do we need the two new variables (`lrcb` and `lrwb`)?

I see them used below, but it's not obvious why we needed these new variables, opposed to using the existing variables like the previous code did; I'm not sure it matters one way or another, but wanted to ask to clear up my understanding.

File: usr/src/uts/common/fs/zfs/zfs_log.c [hunk 3]
```diff
@@ -481,30 +478,26 @@ zfs_log_write(zilog_t *zilog, dmu_tx_t *tx, int txtype,
 	while (resid) {
 		itx_t *itx;
 		lr_write_t *lr;
-		ssize_t len;
+		itx_wr_state_t wr_state = write_state;
```
Comment (amotin) [inline usr/src/uts/common/fs/zfs/zfs_log.c]: The idea was to create local copy for the present chunk to not modify global state if we change it for this for this chunk.  Though changing the global one may also be possible, though I am not sure it would simplify the logic rather then tangling it more.

File: usr/src/uts/common/fs/zfs/zil.c (hunk 2 above)
Comment (amotin) [inline usr/src/uts/common/fs/zfs/zil.c]: One set of variables there points to record in original itx buffer, while another set points to record copy in lwb buffer.  New code logic requires ability to access/modify both of them when splitting single itx record into multiple lwb records.

File: usr/src/uts/common/fs/zfs/zil.c (hunk 1 above)
Comment (amotin) [inline usr/src/uts/common/fs/zfs/zil.c]: "reclen + dlen > lwb_sp"  checks whether record can't fit into lwb completely with all data;
"lwb_sp < ZIL_MAX_LOG_DATA / 8"  checks whether remaining part of lwb significant to bother about record splitting;
"dlen % ZIL_MAX_LOG_DATA == 0 || lwb_sp < reclen + dlen % ZIL_MAX_LOG_DATA"  check whether record split between several lwb's give us residue that could fit nicely into remaining space, not creating extra record/data splits (extra splits require extra zl_get_data() calls).

File: usr/src/uts/common/fs/zfs/zil.c [hunk 4]
```diff
@@ -78,6 +78,13 @@ int zil_replay_disable = 0;
  */
 boolean_t zfs_nocacheflush = B_FALSE;
 
+/*
+ * Limit SLOG write size per commit executed with synchronous priority.
+ * Any writes above that executed with lower (asynchronous) priority to
```
Comment (stevenh) [inline usr/src/uts/common/fs/zfs/zil.c]: Any writes above that *will be* executed with lower...

**Comment by ahrens** [comment]
@zettabot go

File: usr/src/uts/common/fs/zfs/zil.c [hunk 5]
```diff
@@ -78,6 +78,13 @@ int zil_replay_disable = 0;
...
+ * limit potential SLOG device abuse by single active ZIL writer.
+ */
+uint64_t zil_slog_limit = 768 * 1024;
```
Comment (ahrens) [inline usr/src/uts/common/fs/zfs/zil.c]: Why are we changing this from 1MB to 768KB?

File: usr/src/uts/common/fs/zfs/zil.c [hunk 6]
```diff
@@ -908,9 +918,13 @@ zil_lwb_write_init(zilog_t *zilog, lwb_t *lwb)
 		    ZIO_FLAG_CANFAIL);
 	}
 	if (lwb->lwb_zio == NULL) {
+		if (zilog->zl_cur_used <= zil_slog_limit || !lwb->lwb_slog)
+			prio = ZIO_PRIORITY_SYNC_WRITE;
+		else
+			prio = ZIO_PRIORITY_ASYNC_WRITE;
```
Comment (ahrens) [inline usr/src/uts/common/fs/zfs/zil.c]: Async writes are processed in sorted order (by offset), so a steady influx of async writes can starve out some async writes for a long time (imagine one thread that issues a write to the end of the disk, while another thread is writing to every offset starting at the beginning -- the first thread's write will not be processed until the second thread has written the whole disk).  This may not happen commonly for SLOG devices, but it makes it pretty difficult to reason about the performance.

Comment (amotin) [inline usr/src/uts/common/fs/zfs/zil.c]: I understand your worry about possible starvation, but not sure how realistic it is for SLOG.  If somebody have so high permanent overload on SLOG, he should probably consider upgrading hardware or switching to throughput mode.  Another alternative of writing log to main pool is not exactly perfect from both latency and throughput points of view.

**Comment by amotin** [comment]
Any other thoughts about this patch?  I've committed it to FreeBSD two weeks ago.

File: usr/src/uts/common/fs/zfs/zil.c (hunk 6 above)
Comment (richardelling) [inline usr/src/uts/common/fs/zfs/zil.c]: In general, slog writes tend to be ordered with few long seeks, though there can be many short retrogrades.  In any case, today there is no reason to try and optimize slog for HDDs. 

However, I fail to see the advantage here. A slog is its own vdev so it sees no scheduler contention with other I/O types, especially reads. For now, it is not uncommon to need much higher concurrency for SSDs than the default 10, and so it gets tuned [*]. Perhaps a more general solution is to add a separate slog-based priority? In other words, this seems hackish and counter-intuitive to someone trying to tune their system for high throughput with slogs -- a large portion of the installed base including anyone who runs in HA cluster mode.

Similarly, sorting offers no benefit and simply consumes CPU cycles, so taking the async priority to get past the max limit of sync priority seems marginal at best.

Finally, it is not uncommon to see workloads where slogs must be striped to get more throughput and logbias=throughput is not a viable option. We expect this to be fixed in hardware as NVMe/NVMf become more common, but for SAS/SATA/FC we're stuck in a world where the networks are faster than the disk interconnect and disks.

Also, can you explain how this is expected to work with large recordsize?

[*] many flash SSDs cannot reach full write throughput with only 10 concurrent I/Os.

Comment (amotin) [inline usr/src/uts/common/fs/zfs/zil.c]: > Perhaps a more general solution is to add a separate slog-based priority?
Perhaps. My original idea indeed was just to run them with lower priority, but since we have no such mechanism now I've implemented it this way.  I don't mind it to be changed later, it just seems too invasive to do in one patch.

> so taking the async priority to get past the max limit of sync priority seems marginal at best.
The idea was no to get past the limit, but exactly opposite to limit large write burst more strictly to not let it interfere with other really time-critical load.

>Also, can you explain how this is expected to work with large recordsize?
What exactly is your question?  ZIL code has nothing to do with recordsize (aside activation threshold for indirect write, but that is mostly about small recordsize not large). It operates in context of user write operations that can be of any size and alignment.  So even before this patch ZIL code was able to split large write operations between several log records.  The new code just improves space efficiency of this, making for large transactions all log blocks full, except possibly the last one.

**Comment by ahrens** [comment]
@zettabot go

**Comment by ahrens** [comment]
@amotin In terms of changing the semantics of `zil_slog_limit`, I agree that `zl_itx_list_sz` was bogus.  But what about `zl_cur_used`?  It seems like the semantic of "if the current commit is >1MB, don't use SLOG" is potentially useful.  This PR removes that and replaces it with "if the current commit is >768KB, use PRIORITY_ASYNC [but still write to SLOG]".

At a minimum, we should use a different tunable name, to make it clear that the old tunable doesn't exist anymore (there's no way to send ZIL writes to the main pool based on heavy ZIL activity).

I guess I could be convinced that it's unlikely that PRIORITY_ASYNC writes to the SLOG will be starved for a long time.  It still makes me uneasy, but if you are saying that you have tons of real-world experience with this, under heavy ZIL workloads, then I guess it will be OK.

**Comment by amotin** [comment]
The semantic "if the current commit is >1MB, don't use SLOG" could be potentially useful if it would not be so badly broken.  Now, if you have SLOG, *_log_write() functions are never using WR_INDIRECT state, storing all written data into log buffers.  But if there is too much data, then redirecting it to the main pool may not really help but make things worse, causing double write and creating many additional head seeks due to specifics of ZIL allocation.  Actually my changes about postponing log buffer filling could potentially help to fix/improve this by rewriting previously prepared WR_NEED_COPY log records into WR_INDIRECT, but it would require more work and probably made code even more tangled, while the real solution for scenarios of too slow SLOG and big transactions is really logbias=throughput.

If you wish to rename this tunable -- welcome, put there any name you like.  But my personal opinion is that it is not very useful to keep semantics for unknown and possibly undocumented feature that was broken for past 6 years.

What's about starvation, I don't have tons of real-world experience with this so far, but my synthetic tests clearly show that this code does what it should.  In particular, it prevents inadequate bulky load from one misconfigured dataset from badly hurting really latency-critical workload on another dataset.  What's about real world -- this patch is committed to FreeBSD head for more then a month and included into the latest update of FreeNAS released recently.  I don't know how many people who could be affected by this can really diagnose it, but so far everything is quiet.

**Comment by jpaetzel** [comment]
What's the final word on this change?  Matt, if the tunable name is changed are you ok with this going in?  At this point it needs a bit of integration work before it can be merged, but that's ok.  Since it's in FreeBSD now it's causing integration work to be done to ingest OpenZFS commits.

**Comment by ahrens** [comment]
@jpaetzel @amotin I'm OK with the semantics of these changes, given the testing you've done.  Here's what I see left to do to get this integrated:
* rebase on master
* potentially add comments or change some variable names to make the code that @prakashsurya was asking about more clear
* change name of `zil_slog_limit`
* get at least one more detailed code review (potentially from me, I'll also ask around at Delphix - but I can commit to do this in the short time if someone's (Josh or Alexander) still willing to do the other work)

File: usr/src/uts/common/fs/zfs/sys/zil_impl.h [hunk 7]
```diff
@@ -142,6 +141,8 @@ typedef struct zil_bp_node {
 
 #define	ZIL_MAX_LOG_DATA (SPA_OLD_MAXBLOCKSIZE - sizeof (zil_chain_t) - \
     sizeof (lr_write_t))
+#define	ZIL_MAX_COPIED_DATA \
+    ((SPA_OLD_MAXBLOCKSIZE - sizeof (zil_chain_t)) / 2 - sizeof (lr_write_t))
```
Comment (amotin) [inline usr/src/uts/common/fs/zfs/sys/zil_impl.h]: This maximum value is based on assumption that we should fit at least two log records into single maximum-sized log block.  Such block has maximal size of SPA_OLD_MAXBLOCKSIZE, and includes zil_chain_t header per block, and lr_write_t header per (each of two) record.  If data block is bigger then this value, then we should not assume we'll be lucky to fit it, and should postpone the copying till later stage, when it can be done more efficiently.

**Comment by amotin** [comment]
> Do you have any data on the latencies from the performance tests you mentioned?

I did plenty of measurements, but it was six months ago.  My tests did confirmed that bulk log writer indeed reduced its effects on really synchronous writer.  I don't have those numbers now, but I measured much better IOPS for short synchronous latency-critical writer, while doing large commits to another dataset at full SLOG device throughput.  I understand described worry about potentially unbound wait, but as I have already said, it can be fixed separately.

What's about general ZIL writing efficiency improvements, here is some numbers from those old tests:
Scenario 1 (pool of single HDD, measuring throughput of synchronous writes of different sizes to dataset with 128KB records):

              Before   After
    64K         5M      5M
    128K        5M      7M
    256K       10M      9.5M
    512K       13M     13M
    1M         18M     20M
    2M         21M     26M
    4M         24M     35M
    8M         24M     43M

Performance improved here (especially for large commits) due to more efficient data placement in log blocks, and so lower number of seeks and minimal useless head fly time over allocated but unused surface areas.

Scenario 2 (pool of 12 SSDs with NVDIMM SLOG, measuring throughput of synchronous writes of different sizes to dataset with 128KB records):

            Before    After   With other ZIO shortcuts
    64K       284M     294M     768M
    128K      557M     507M     513M
    256K      817M     784M
    512K     1002M    1017M
    1M       1162M    1163M
    2M       1188M    1288M
    4M       1222M    1350M
    8M       1144M    1385M

Performance of large commits improved due to smaller number of log I/Os to process.  But main benefit invisible here is twice lower SLOG space usage, which is more expensive and limited for NVDIMMs then for SSDs.

I tested many other scenarios too, but general tendencies are the same as shown here -- smaller number of ZIL write operations reduce head seek times and processing overhead, better space utilization improves throughput for HDDs and space efficiency for SSDs/NVDIMMs.

File: usr/src/uts/common/fs/zfs/sys/zil_impl.h (hunk 7 above)
Comment (brad-lewis) [inline usr/src/uts/common/fs/zfs/sys/zil_impl.h]: That makes sense.  I think a little comment here would have put me on track to figure it out: 
/* Fit at least two log records into each maximum sized block */

File: usr/src/uts/common/fs/zfs/zil.c (hunk 1 above)
Comment (brad-lewis) [inline usr/src/uts/common/fs/zfs/zil.c]: If the first clause was changed to "reclen >= lwb_sp" would that eliminate the need for   "dlen % ZIL_MAX_LOG_DATA == 0"?

File: usr/src/uts/common/fs/zfs/zil.c [hunk 8]
```diff
@@ -1107,16 +1117,19 @@ zil_lwb_commit(zilog_t *zilog, itx_t *itx, lwb_t *lwb)
...
-			if (dlen) {
-				ASSERT(itx->itx_wr_state == WR_NEED_COPY);
+			if (itx->itx_wr_state == WR_NEED_COPY) {
 				dbuf = lr_buf + reclen;
-				lrw->lr_common.lrc_reclen += dlen;
+				lrcb->lrc_reclen += dnow;
+				if (lrwb->lr_length > dnow)
+					lrwb->lr_length = dnow;
+				lrw->lr_offset += dnow;
+				lrw->lr_length -= dnow;
```
Comment (brad-lewis) [inline usr/src/uts/common/fs/zfs/zil.c]: It would be much clearer and easier to maintain if local variables kept the updated offset and length values.  This block would be only setting values of the itx in the buffer.

**Comment by brad-lewis** [comment]
Your numbers looks great, thanks.  I think there is much goodness here, especially in the space efficiency improvements.   I commented with a couple of suggestions to make the zil_lwb_commit changes clearer.

File: usr/src/uts/common/fs/zfs/zil.c (hunk 8 above)
Comment (amotin) [inline usr/src/uts/common/fs/zfs/zil.c]: May be, but I don't want to change the logic of the code much at this point without significant reason.  This code is already committed to FreeBSD and diverging from it makes testing difficult for me.  Lets do it this "good" way which was well tested, remove divergence and then look for "better".

**Comment by amotin** [comment]
I added some comments as requested.

**Comment by ahrens** [comment]
@zettabot go

File: usr/src/uts/common/fs/zfs/zil.c (hunk 8 above)
Comment (brad-lewis) [inline usr/src/uts/common/fs/zfs/zil.c]: I understand the reluctance to make changes and retest.  I think this review is dragging on because the code, especially in zil_lwb_commit() is difficult to understand raising concerns about being able to maintain and improve on the code in the future.  The code here to manage splitting the blocks is pretty similar to that in zfs_log_write().  The loop and use of local offset and resid variables make it much more obvious what is going on.  Perhaps it would be worth it to rewrite zil_lwb_commit with the inner part to something like this: 

`static lwb_t *
zil_lwb_commit(zilog_t *zilog, itx_t *itx, lwb_t *lwb)
{
        lr_t *lrc = &itx->itx_lr; /* common log record */
        lr_write_t *lrw = (lr_write_t *)lrc;
        lr_t *lrc_orig = lrc;
        uint64_t offset = lrw->lr_offset;
        uint64_t length = lrw->lr_length;
        char *lr_buf;
        uint64_t txg = lrc->lrc_txg;
        uint64_t reclen = lrc->lrc_reclen;
        uint64_t dlen = 0;
        uint64_t dnow;
        uint64_t lwb_sp;

        if (lwb == NULL)
                return (NULL);

        ASSERT(lwb->lwb_buf != NULL);

        if (lrc->lrc_txtype == TX_WRITE && itx->itx_wr_state == WR_NEED_COPY)
                dlen = P2ROUNDUP_TYPED(
                    lrw->lr_length, sizeof (uint64_t), uint64_t);

        zilog->zl_cur_used += dlen;

        zil_lwb_write_init(zilog, lwb);

        while (dlen > 0 || lrc == lrc_orig)  {
                zilog->zl_cur_used += reclen;
                /*
                 * If this record won't fit in the current log block,
                 * start a new one.  For WR_NEED_COPY optimize layout
                 * for minimal number of chunks.
                 */
                lwb_sp = lwb->lwb_sz - lwb->lwb_nused;
                if (reclen > lwb_sp || (reclen + dlen > lwb_sp &&
                    lwb_sp < ZIL_MAX_WASTE_SPACE && (dlen % ZIL_MAX_LOG_DATA
                    == 0 || lwb_sp < reclen + dlen % ZIL_MAX_LOG_DATA))) {
                        lwb = zil_lwb_write_start(zilog, lwb);
                        if (lwb == NULL)
                                return (NULL);
                        zil_lwb_write_init(zilog, lwb);
                        ASSERT(LWB_EMPTY(lwb));
                        lwb_sp = lwb->lwb_sz - lwb->lwb_nused;
                        ASSERT3U(reclen + MIN(dlen, sizeof(uint64_t)),
                            <=, lwb_sp);
                }

                dnow = MIN(dlen, lwb_sp - reclen);
                lr_buf = lwb->lwb_buf + lwb->lwb_nused;
                bcopy(lrc_orig, lr_buf, reclen);
                lrc = (lr_t *)lr_buf;
                lrw = (lr_write_t *)lrc;
                /*
                 * If it's a write, fetch the data or get its blkptr
                 * as appropriate.
                 */
                if (lrc->lrc_txtype == TX_WRITE) {
                        if (txg > spa_freeze_txg(zilog->zl_spa))
                                txg_wait_synced(zilog->zl_dmu_pool, txg);
                        if (itx->itx_wr_state != WR_COPIED) {
                                char *dbuf;
                                int error;

                                if (itx->itx_wr_state == WR_NEED_COPY) {
                                        dbuf = lr_buf + reclen;
                                        lrc->lrc_reclen += dnow;
                                        lrw->lr_length = length;
                                        lrw->lr_offset = offset;
                                        offset += dnow;
                                        length -= dnow;
                                } else {
                                        ASSERT(itx->itx_wr_state == WR_INDIRECT);
                                        dbuf = NULL;
                                }
                                error = zilog->zl_get_data(
                                    itx->itx_private, lrw, dbuf, lwb->lwb_zio);
                                if (error == EIO) {
                                        txg_wait_synced(zilog->zl_dmu_pool, txg);
                                        return (lwb);
                                }
                                if (error != 0) {
                                        ASSERT(error == ENOENT ||
                                            error == EEXIST ||
                                            error == EALREADY);
                                        return (lwb);
                                }
                        }
                }

                /*
                 * We're actually making an entry, so update lrc_seq to be the
                 * log record sequence number.  Note that this is generally not
                 * equal to the itx sequence number because not all transactions
                 * are synchronous, and sometimes spa_sync() gets there first.
                 */
                lrc->lrc_seq = ++zilog->zl_lr_seq; /* we are single threaded */
                lwb->lwb_nused += reclen + dnow;
                lwb->lwb_max_txg = MAX(lwb->lwb_max_txg, txg);
                ASSERT3U(lwb->lwb_nused, <=, lwb->lwb_sz);
                ASSERT0(P2PHASE(lwb->lwb_nused, sizeof (uint64_t)));

                dlen -= dnow;
        }

        return (lwb);
}
`

**Comment by ahrens** [comment]
@amotin I'm good with integrating this once you make the changes Brad suggested https://github.com/openzfs/openzfs/pull/219#discussion_r111232889

File: usr/src/uts/common/fs/zfs/zil.c (hunk 8 above)
Comment (amotin) [inline usr/src/uts/common/fs/zfs/zil.c]: I agree that this code is not trivial, but I don't think this change simplify it significantly.  The idea of substituting lrc/lrw variables meaning and resulted existence of lrc_orig variable is no less confusing to me.

File: usr/src/uts/common/fs/zfs/zil.c [hunk 9]
```diff
@@ -908,9 +918,13 @@ zil_lwb_write_init(zilog_t *zilog, lwb_t *lwb)
 		    ZIO_FLAG_CANFAIL);
 	}
 	if (lwb->lwb_zio == NULL) {
+		if (zilog->zl_cur_used <= zil_slog_bulk || !lwb->lwb_slog)
```
Comment (avg-I) [inline usr/src/uts/common/fs/zfs/zil.c]: I would put the `lwb_slog` check first, so that it is more clear that we do not care about the used size when we are not using slog.

File: usr/src/uts/common/fs/zfs/zil.c (hunk 2 above)
Comment (avg-I) [inline usr/src/uts/common/fs/zfs/zil.c]: I would either split the declaration lines here so that there is one variable per declaration or I would move the initializations to separate lines. It is harder to read the code when there are both initialized and uninitialized variables in the same declarations.
Also, maybe it would make sense to add your description of the variables as a comment in the code.

File: usr/src/uts/common/fs/zfs/zio.c [hunk 10]
```diff
@@ -2860,25 +2860,24 @@ zio_dva_unallocate(zio_t *zio, zio_gang_node_t *gn, blkptr_t *bp)
...
 {
 	int error = 1;
 	zio_alloc_list_t io_alloc_list;
 
 	ASSERT(txg > spa_syncing_txg(spa));
 
 	metaslab_trace_init(&io_alloc_list);
-
-	if (use_slog) {
-		error = metaslab_alloc(spa, spa_log_class(spa), size,
```
Comment (avg-I) [inline usr/src/uts/common/fs/zfs/zio.c]: I am not sure, but maybe it would be better to use the input value of `slog` to decide if the `spa_log_class` allocation attempt should be skipped.  I suggest this only as a micro-optimization to avoid calling `metaslab_alloc` and `spa_config_enter` / `spa_config_exit` in it when there is no slog device.

File: usr/src/uts/common/fs/zfs/zvol.c [hunk 11]
```diff
@@ -1058,54 +1058,44 @@ zvol_log_write(zvol_state_t *zv, dmu_tx_t *tx, offset_t off, ssize_t resid,
 {
```
Comment (avg-I) [inline usr/src/uts/common/fs/zfs/zvol.c]: I wonder if it would be possible to share some code between this function and `zfs_log_write`.
Not necessary for this change, but would be nice to do in the future.

Comment (amotin) [inline usr/src/uts/common/fs/zfs/zvol.c]: It would probably be nice, but so far I just tried to unify them.

File: usr/src/uts/common/fs/zfs/zio.c (hunk 10 above)
Comment (amotin) [inline usr/src/uts/common/fs/zfs/zio.c]: There is no input value for slog parameter any more.  It is write-only.

File: usr/src/uts/common/fs/zfs/zil.c (hunk 9 above)
Comment (amotin) [inline usr/src/uts/common/fs/zfs/zil.c]: Agreed. Changed.

File: usr/src/uts/common/fs/zfs/zil.c (hunk 2 above)
Comment (amotin) [inline usr/src/uts/common/fs/zfs/zil.c]: OK.  I moved initialization on separate lines and added some comments.

File: usr/src/uts/common/fs/zfs/zil.c (hunk 8 above)
Comment (stevenh) [inline usr/src/uts/common/fs/zfs/zil.c]: I'd have to agree with @amotin here, while this could do with cleaning up, so its easier to understand, I believe this would be best done as a separate PR.

**Comment by amotin** [comment]
So, if there are no objections any more, can this finally be committed to close this chapter?  The life is goes on, and I see another potentially conflicting ZIL change being proposed for discussion.

**Comment by ahrens** [comment]
@amotin Yes, we will commit this change soon - at least by the end of the week if not sooner.

**Comment by prakashsurya** [comment]
landed via https://github.com/openzfs/openzfs/commit/aeb13accf5ba027cc5d351fdd5f4e9a6fa905431

**Comment by prakashsurya** [comment]
thank you for your patience, @amotin !
